### PR TITLE
feat: export raw code coverage data from instrumented function handlers

### PR DESCRIPTION
**Pull request by mickdekkers**
The code coverage data will be saved in json format at the path specified
in the `SLS_OFFLINE_COVERAGE_DATA_OUTPUT_PATH` environment variable.

For this to work, the function handler code needs to be instrumented using
something like [babel-plugin-istanbul](https://github.com/istanbuljs/babel-plugin-istanbul).

Note that this outputs _raw_ coverage data. Reports can be generated from this
data in various formats (see [istanbul-reports/lib](https://github.com/istanbuljs/istanbuljs/tree/master/packages/istanbul-reports/lib)) using istanbul's internal packages.

Example reporting script:
```js
const path = require('path');
const libReport = require('istanbul-lib-report');
const libCoverage = require('istanbul-lib-coverage');
const HtmlReport = require('istanbul-reports/lib/html');

const rawCoverage = require('./YOUR_COVERAGE_DATA.json');

const context = libReport.createContext({
    dir: path.resolve(__dirname, 'coverage'), // this is the report output directory
});

const tree = libReport.summarizers.pkg(
    libCoverage.createCoverageMap(rawCoverage),
);

const report = new HtmlReport({}); // (the empty object is required)

tree.visit(report, context);

// The HTML report is now saved to the output directory
```

### CODE REVIEW

File: src/index.js [hunk 1]
```diff
@@ -39,6 +46,55 @@ class Offline {
...
+    this.shouldOutputCoverage = typeof this.coverageDataOutputPath !== 'undefined';
+
+    if (this.shouldOutputCoverage) {
+      debugLog(
+        `Code coverage will be written to ${this.coverageDataOutputPath}`
+      );
+      /* eslint-disable new-cap */
+      this.coverageQueue = new pQueue({ concurrency: 1 });
+      /* eslint-enable new-cap */
+      this.outputCoverage = pDebounce(
```
Comment (mickdekkers) [inline src/index.js:57]: The coverage output code is [debounced](https://css-tricks.com/debouncing-throttling-explained-examples/#article-header-id-0) so it doesn't spam I/O while tests are being executed in rapid succession. The queue ensures the output file is not written to multiple times at once.

I'm not sure the queue is 100% necessary, but it covers the edge case where a file write takes longer than the debounce duration. I guess it might apply when saving code coverage for very large lambda functions or on a very slow disk.

File: src/index.js [hunk 2]
```diff
@@ -39,6 +46,55 @@ class Offline {
...
+              const summary = coverageMap
+                .files()
+                .reduce(
+                  (sum, f) => sum.merge(coverageMap.fileCoverageFor(f).toSummary()),
+                  libCoverage.createCoverageSummary(),
+                );
+              debugLog('Global coverage summary', summary);
+            }
+          }),
+        // TODO: perhaps make this configurable
```
Comment (mickdekkers) [inline src/index.js]: Thoughts? This value is the debounce duration.

Comment (wouterb) [reply src/index.js]: You could move it to a constant, to make clarify this magic number. I do not foresee an usecase for changing this value.

**Review by wouterb — APPROVED** [review]

File: src/index.js [hunk 3]
```diff
@@ -39,6 +46,55 @@ class Offline {
...
     this.provider = 'aws';
     this.start = this.start.bind(this);
+    this.coverageDataOutputPath = process.env.SLS_OFFLINE_COVERAGE_DATA_OUTPUT_PATH;
+    this.shouldOutputCoverage = typeof this.coverageDataOutputPath !== 'undefined';
+
+    if (this.shouldOutputCoverage) {
+      debugLog(
+        `Code coverage will be written to ${this.coverageDataOutputPath}`
+      );
+      /* eslint-disable new-cap */
```
Comment (M-TGH) [inline src/index.js]: Can be changed to a single line disable:
```suggestion
      /* eslint-disable-next-line new-cap */
```

File: src/index.js [hunk 4]
```diff
@@ -39,6 +46,55 @@ class Offline {
...
+      this.coverageQueue = new pQueue({ concurrency: 1 });
+      /* eslint-enable new-cap */
```
Comment (M-TGH) [inline src/index.js]: No need to re-enable if you implement other suggestion. (for applying bulk suggestions):
```suggestion
```

File: src/index.js (hunk 3 above)
Comment (mickdekkers) [reply src/index.js]: I tried this initially, but for some reason eslint isn't picking up the single line disables in this repo... Might just be on my end. Could you try to reproduce this?

File: src/index.js (hunk 2 above)
Comment (mickdekkers) [reply src/index.js]: Done 👍

File: src/index.js (hunk 3 above)
Comment (M-TGH) [reply src/index.js]: Ah, rather old version of eslint, could use `eslint-disable-line` on the line in question.

File: package.json [hunk 5]
```diff
@@ -123,7 +123,12 @@
     "jsonpath-plus": "^0.16.0",
     "jsonschema": "^1.2.4",
     "jsonwebtoken": "^8.5.0",
+    "lodash": "^4.17.10",
```
Comment (wouterb) [inline package.json:126]: Looks like this one is not used.
```suggestion
```

File: package.json [hunk 6]
```diff
@@ -1,5 +1,5 @@
 {
-  "name": "serverless-offline",
+  "name": "@tradecast/serverless-offline",
```
Comment (wouterb) [inline package.json:2]: publishing to our `@tradecast` scope requires an publishConfig:
```suggestion
  "name": "@tradecast/serverless-offline",
  "publishConfig": {
        "always-auth": true,
        "registry": "https://npm.tradecast.eu:4879"
    },
```

**Review by wouterb — CHANGES_REQUESTED** [review]

**Comment by richardcornelissen** [comment]
Moved to #2
